docs(feature/pace-51): udf tutorial (#81)
getstrm/pace: docs/tutorials/udfs.md

---

> description: Create and use a Python UDF on Databricks.

Reviewer: ```suggestion
description: Create and use a Python UDF on Databricks
```

---
# User Defined Functions in Python
In this tutorial we will

> * create a User Defined Function written in Python. We will use DataBricks as a processing platform.

Reviewer: ```suggestion
* create a User Defined Function written in Python. We will use Databricks as a processing platform.
```

* create a Data Policy that uses this UDF, and apply it to the the platform.
* inspect the resulting SQL View.
* show some results.

[databricks-udf]: https://docs.databricks.com/en/udf/python.html
## Creating the UDF
On Databricks this is [quite simple][databricks-udf]. In SQL we would do the following.
Make sure we execute it in the catalog and schema of the data.
```sql
CREATE OR REPLACE FUNCTION squareWithPython(age INT)
RETURNS INT
LANGUAGE PYTHON
AS $$
return age*age
$$;
```
NOTE: Make sure both the PACE service credential, as well as any user that might access the resulting SQL VIEW
on Databricks has `EXECUTE` permissions on the function.
![exec permissions](./dbr-exec-perms.png)

> ## Create a data-policy

Reviewer: ```suggestion
## Create a Data Policy
```

> I've used a demo table with an `age` integer column in it, and downloaded a blueprint data policy

Reviewer: ```suggestion
We have used a demo table with an `age` integer column in it, and downloaded a blueprint data policy:
```

>     pace  get data-policy --processing-platform dbr-pace pace.alpha_test.gddemo > policy.yaml

Reviewer: Shall we use just `demo` instead of `gddemo`?

Author: is goed.

> I've edited the policy file, and included the following field transformation.

Reviewer: ```suggestion
We have then edited the policy file, and included the following field transformation:
```

```yaml
metadata:
  ...
  version: 1
platform:
  id: dbr-pace
  platform_type: DATABRICKS
rule_sets:
- field_transforms:
  - field:
      name_parts: [ age ]
    transforms:
      - sql_statement: 
          statement: pace.alpha_test.squarewithpython(age)
        principals: []
  target:
    fullname: pace.alpha_test.gddemo_pace_view
source:
  ...
  ref: pace.alpha_test.gddemo
```
So this field transformation defines that any user (`principals: []`) receives the squared value of the `age` column.

> ## Create the SQL VIEW on Databricks:

Reviewer: ```suggestion
## Create the SQL VIEW on Databricks
```

First `upsert` the policy file to PACE.

    pace upsert data-policy policy.yaml

> And the actually apply it on the processing platform.

Reviewer: ```suggestion
And then actually apply it on the processing platform.
```

    pace apply data-policy pace.alpha_test.gddemo  --processing-platform dbr-pace 

> ## Investigate the results.

Reviewer: ```suggestion
## Investigate the results
```

If everything went right we would have a SQL VIEW named `pace.alpha_test.gddemo_pace_view` with this
view definition:

```sql
-- VIEW DEFINITION
select
    transactionId, userId, email, pace.alpha_test.squarewithpython(age) age,
    size, hairColor, transactionAmount, items, itemCount, date, purpose
    from pace.alpha_test.gddemo
```

The original `gdddemo` table contains fairly normal ages:
```text
select email, age from pace.alpha_test.gddemo limit 5;
email                     age
jeffreypowell@hotmail.com 33
forbeserik@gmail.com      16
wboone@gmail.com          64
oliverjulie@yahoo.com     12
debra64@hotmail.com       79
```

> But the VIEW not so much :-)

Reviewer: ```suggestion
But the _view_ clearly shows them squared:
```

Author: boring :-)
